README.md
align
=====
**align** **-C**|**c**|**R**|**r** [**-w** *width*]

Copies a file from standard input to standard output, indented with
spaces to be centered or right-align, depending on the chosen mode:

 - **-C**: Center as a block
 - **-c**: Center per line
 - **-R**: Right-align as a block
 - **-r**: Right-align per line

Tabs are expanded. The target width can be set with **-w** *width* and
defaults to the COLUMNS environment variable, or 80 if unset.

Examples
--------
Centering a file in block and per-line modes respectively:

    $ align -C <poem.txt

           Into the machine
           The soul descended
           Speaking to its creator
           Greeting the world

    $ align -c <poem.txt

              Into the machine
             The soul descended
           Speaking to its creator
             Greeting the world

Building
--------
Should work on Unix (Linux, BSD, macOS, ...) or on Windows with
minor changes:

    $ make
    $ sudo make install

Author
------
By Sijmen J. Mulder (<ik@sjmulder.nl>), June 2020. See LICENSE.md.


CHANGES.md
align 1.1.1
-----------
11 January 2020

 - Fixed length calculation for lines with leading tabs.
 - Now using MANPREFIX convention in Makefile.

align 1.1
---------
5 June 2020

Now with 4 modes: center block, center lines, right-align block,
right-align lines. Name changed from *center* to *align* te reflect
the new functionality.

center 1.0
----------
3 June 2020

Initial release.


align.c
/* Copyright (c) 2020, Sijmen J. Mulder (see LICENSE.md) */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <err.h>

static FILE *copytmp(FILE *);
static int getlnwidth(FILE *);

/* Mode, e.g. M_CENTER | M_BLOCK */
#define M_CENTER 0x01
#define M_RIGHT  0x02
#define M_BLOCK  0x10
#define M_LINE   0x20

int
main(int argc, char *argv[])
{
	int c, n, i;
	int inwidth=0, outwidth=-1;
	int mode=0;
	int npad, fst=1;
	char *str;
	FILE *input = stdin;
	fpos_t pos;

#if __OpenBSD__
	if (pledge("stdio tmppath", NULL) == -1)
		err(1, "pledge");
#endif

	while ((c = getopt(argc, argv, "CcRrw:")) != -1)
		switch (c) {
			case 'C': mode = M_CENTER | M_BLOCK; break;
			case 'c': mode = M_CENTER | M_LINE;  break;
			case 'R': mode = M_RIGHT  | M_BLOCK; break;
			case 'r': mode = M_RIGHT  | M_LINE;  break;
			case 'w': outwidth = atoi(optarg); break;
			default: return -1;
		}

	if (!mode || argc > optind)
		errx(1, "usage: align -C|c|R|r [-w width] <file");

	/* Use $COLUMN as default with, or 80 if unset */
	if (outwidth < 0 && (str = getenv("COLUMNS")))
		outwidth = strtol(str, NULL, 10);
	if (outwidth < 0)
		outwidth = 80;

	/* We must be able to seek */
	if (fseek(input, SEEK_SET, 0) == -1)
		input = copytmp(input);

	if (mode & M_BLOCK) {
		while (!feof(input))
			if ((n = getlnwidth(input)) > inwidth)
				inwidth = n;
		if (fseek(input, SEEK_SET, 0) == -1)
			err(1, NULL);
	}

	while ((c = fgetc(input)) != EOF) {
		if (fst) {
			if (mode & M_LINE) {
				if (fgetpos(input, &pos) == -1)
					err(1, NULL);

				ungetc(c, input);
				inwidth = getlnwidth(input);

				if (fsetpos(input, &pos) == -1)
					err(1, NULL);
			}

			if ((npad = outwidth-inwidth) > 0) {
				if (mode & M_CENTER)
					npad = (npad+1)/2;
				for (i = 0; i < npad; i++)
					fputc(' ', stdout);
			}

			fst = 0;
		}

		switch (c) {
			case '\t': fputs("        ", stdout); break;
			case '\n': fputc(c, stdout); fst=1; break;
			default:   fputc(c, stdout); break;
		}
	}

	return 0;
}

/* Makes a copy of the stream to an unnamed temporary file */
static FILE *
copytmp(FILE *f)
{
	static char buf[4096];
	FILE *tmp;
	size_t nr;

	if (!(tmp = tmpfile()))
		err(1, NULL);
	while ((nr = fread(buf, 1, sizeof(buf), f)))
		if (fwrite(buf, 1, nr, tmp) != nr)
			err(1, NULL);
	if (ferror(f))
		err(1, NULL);
	if (fseek(tmp, SEEK_SET, 0) == -1)
		err(1, NULL);

	return tmp;
}

/* Scans a line and returns its width */
static int
getlnwidth(FILE *f)
{
	int w=0, c;

	while ((c = fgetc(f)) != EOF && c != '\n')
		switch (c) {
			case '\t': w = (w+8)/8 * 8; break;
			default: w++; break;
		}
	if (ferror(f))
		err(1, NULL);

	return w;
}


center.c
/* Copyright (c) 2020, Sijmen J. Mulder (see LICENSE.md) */

#include <stdio.h>
#include <stdlib.h>
#include <unicode/ustdio.h>
#include <unicode/uchar.h>
#include <errno.h>
#include <unistd.h>
#include <err.h>

static FILE *copytmp(FILE *);
static int maxwidth(UFILE *);
static void center(UFILE *, UFILE *, int, int);

int
main(int argc, char *argv[])
{
	int c;
	int inwidth;
	int outwidth = -1;
	char *str;
	FILE *input = stdin;
	UFILE *uin, *uout;

#if __OpenBSD__
	if (pledge("stdio tmppath", NULL) == -1)
		err(1, "pledge");
#endif

	while ((c = getopt(argc, argv, "w:")) != -1)
		switch (c) {
			case 'w': outwidth = atoi(optarg); break;
			default: return -1;
		}

	if (argc > optind) {
		fprintf(stderr, "usage: center [-w width]\n");
		return -1;
	}

	/*
	 * Use $COLUMNS as terminal width if no -w given, or 80 if
	 * unset.
	 */
	if (outwidth < 0 && (str = getenv("COLUMNS")))
		outwidth = strtol(str, NULL, 10);
	if (outwidth < 0)
		outwidth = 80;

	/*
	 * We're making 2 passes through the file, first to measure
	 * line lengths, then for the output, so we must be able to
	 * rewind. Test that. If we can't, copy the stream to a
	 * temporary file.
	 */
	if (fseek(input, SEEK_SET, 0) == -1)
		input = copytmp(input);

	if (!(uin = u_finit(input, NULL, NULL)))
		errx(1, "u_finit(input) failed");
	if (!(uout = u_finit(stdout, NULL, NULL)))
		errx(1, "u_finit(stdout) failed");

	inwidth = maxwidth(uin);
	if (fseek(input, SEEK_SET, 0) == -1)
		err(1, NULL);

	center(uin, uout, inwidth, outwidth);

	return 0;
}

/*
 * Copies the given file stream to a temporary, unlinked file and
 * returns a handle to that file at pos 0.
 */
static FILE *
copytmp(FILE *f)
{
	static char buf[4096];
	FILE *tmp;
	size_t nr;

	if (!(tmp = tmpfile()))
		err(1, NULL);
	while ((nr = fread(buf, 1, sizeof(buf), f)))
		if (fwrite(buf, 1, nr, tmp) != nr)
			err(1, NULL);
	if (ferror(f))
		err(1, NULL);
	if (fseek(tmp, SEEK_SET, 0) == -1)
		err(1, NULL);

	return tmp;
}

/*
 * Goes through the given file, starting at the current position, and
 * returns the character count of the longest line. Every grapheme
 * is considered a single character. Does not rewind the file.
 */
static int
maxwidth(UFILE *f)
{
	int w=0, mw=0;
	UChar32 c;

	while ((c = u_fgetcx(f)) != U_EOF)
		if (c == '\n') {
			if (w > mw)
				mw = w;
			w = 0;
		} else if (c == '\t')
			w = (w+8)/8 * 8;
		else if (u_hasBinaryProperty(c, UCHAR_GRAPHEME_BASE))
			w++;

	return mw;
}

/*
 * Goes through the given input file and copies it to the given output
 * file using spaces for indentation to center the text of width
 * inwidth in a container of outwidth. Tabs are expanded.
 */
static void
center(UFILE *in, UFILE *out, int inwidth, int outwidth)
{
	int i;
	UChar32 c=0;

	while ((c = u_fgetcx(in)) != U_EOF) {
		if (c == '\n') {
			u_fputc('\n', out);
			continue;
		}

		for (i = 0; i < (outwidth-inwidth)/2; i++)
			u_fputc(' ', out);
		do {
			if (c == '\t')
				u_fprintf(out, "        ");
			else
				u_fputc(c, out);
		} while (c != '\n' && (c = u_fgetcx(in)) != EOF);
	}
}
